### PR TITLE
sourcery: init at 1.15.0

### LINKED ISSUE #243774: Package request: python-sourcery + VS Code extension
**Project description**

Automatically review and improve your Python code.

**Metadata**

* homepage URL: https://sourcery.ai/
* source URL: https://github.com/sourcery-ai/sourcery
* license: mit
* platforms: unix, linux, darwin

**Analysis**

Naively installing the package with python works, but at runtime this error is shown

```
/home/fps/lg/.venv/lib/python3.10/site-packages/sourcery/sourcery: error while loading shared libraries: libcrypt.so.1: cannot open shared object file: No such file or directory
```

most likely because it is a generic linux binary and we need to patchelf it. Is this analysis correct?

**Pointers**

Ideally the associated VS Code [extension](https://marketplace.visualstudio.com/items?itemName=sourcery.sourcery) should be available as well. Related PR is [#141772](https://github.com/NixOS/nixpkgs/pull/141772).

### LINKED ISSUE #244143: Package request: Sourcery.ai
**Add Sourcery.ai to packages**

Sourcery.ai is a non free package for Python which is available from this [link](https://pypi.org/project/sourcery/).
Or it can be made available for VScode via this [link](https://marketplace.visualstudio.com/items?itemName=sourcery.sourcery).
Similar for Vim, Command Line, PyCharm, Emacs or Sublime.

I tried to add this this to my home.nix within a vsocde section:
      {
       name = "sourcery.sourcery";
        publisher = "sourcery.ai";
        version = "1.6.0";
        sha256 = "0000000000000000000000000000000000000000000000000000";
      }

But result:
Cant install sourcery for vscode:
home-manager build switch
produces
building '/nix/store/ag5i8qpv0wkpayjvnkdi0dnwdmhrqa12-sourcery.ai-sourcery.sourcery.zip.drv'...

trying https://sourcery.ai.gallery.vsassets.io/_apis/public/gallery/publisher/sourcery.ai/extension/sourcery.sourcery/1.6.0/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 400
error: cannot download sourcery.ai-sourcery.sourcery.zip from any mirror
error: builder for '/nix/store/ag5i8qpv0wkpayjvnkdi0dnwdmhrqa12-sourcery.ai-sourcery.sourcery.zip.drv' failed with exit code 1;
       last 7 log lines:
       >
       > trying https://sourcery.ai.gallery.vsassets.io/_apis/public/gallery/publisher/sourcery.ai/extension/sourcery.sourcery/1.6.0/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage
       >   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
       >                                  Dload  Upload   Total   Spent    Left  Speed
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       > curl: (22) The requested URL returned error: 400
       > error: cannot download sourcery.ai-sourcery.sourcery.zip from any mirror
       For full logs, run 'nix log /nix/store/ag5i8qpv0wkpayjvnkdi0dnwdmhrqa12-sourcery.ai-sourcery.sourcery.zip.drv'.
error: 1 dependencies of derivation '/nix/store/z62dqgjxbm3v3p4hj62bf4aa9nv8y1ix-vscode-extension-sourcery.ai-sourcery.sourcery-1.6.0.drv' failed to build
error: 1 dependencies of derivation '/nix/store/y3rmcb7pcc9j03j0408dsn6g5cnzkvw5-hm_.vscodeextensions.extensionsimmutable.json.drv' failed to build
error: 1 dependencies of derivation '/nix/store/ixfc6bjhprz4n88zk6rsgzs8kas925xi-home-manager-files.drv' failed to build
error: 1 dependencies of derivation '/nix/store/qfxzn9qqbzpq5mmsk1py6ni9lw70zhgk-home-manager-generation.drv' failed to build`

When I visit the link in browser I get:
`Our services aren't available right now`

I note that searching Nix packages for sourcery produces a result for Code Sourcery C++ ABI, which is unrelated, hence the name should be distinctive.

**Metadata**

* homepage URL: https://docs.sourcery.ai/
* source URL: https://github.com/sourcery-ai/sourcery
and https://www.vsixhub.com/vsix/20643/
and https://api.github.com/repos/sourcery-ai/sourcery/tarball/v1.6.0
and https://api.github.com/repos/sourcery-ai/sourcery/zipball/v1.6.0
* license: mit
* platforms: linux, darwin

### PR DESCRIPTION
**Pull request by TomaSajt**
###### Description of changes

Closes: https://github.com/NixOS/nixpkgs/issues/243774
Closes: https://github.com/NixOS/nixpkgs/issues/244143

This PR adds the Sourcery.ai Python package and the VSCode extension.
Although the repo says it's MIT licensed, the Pypi pacakge says it's proprietary (which it actually is).

For the Python package, there are only platform specific wheels available on Pypi.
I don't know which platforms are supported actually, so I'll just go with x86_64 for linux and darwin.

The VSCode extension has all platforms' sources in it, so hopefully that will work as expected.

https://github.com/NixOS/nixpkgs/issues/243774 said that the package name was `python-sourcery`, but that's another, unrelated package

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jfvillablanca** [comment]
Result of `nixpkgs-review pr 244908` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>python310Packages.sourcery</li>
    <li>python310Packages.sourcery.dist</li>
    <li>python311Packages.sourcery</li>
    <li>python311Packages.sourcery.dist</li>
    <li>vscode-extensions.sourcery.sourcery</li>
  </ul>
</details>

**Review by MaxwellDupre — APPROVED** [review]
LGTM, except more likely to get merged if only one commit.

**Comment by TomaSajt** [comment]
Well, it's been a while since I took a better look at this PR.
I rebased to a more recent commit.
I updated the version to `1.15.0`
I moved the package to `pkgs/by-name` while using `buildPythonApplication` instead of treating this as a normal python package (because it is not).
I also combined the two commits into one.

**Comment by TomaSajt** [comment]
@ofborg build sourcery

**Comment by TomaSajt** [comment]
@ofborg build sourcery

**Review by MaxwellDupre — APPROVED** [review]
Looks good.

File: pkgs/applications/editors/vscode/extensions/default.nix [hunk 1]
```diff
@@ -23,6 +23,7 @@
 , typst-preview
 , autoPatchelfHook
 , zlib
+, libxcrypt-legacy
```
Comment (superherointj) [inline pkgs/applications/editors/vscode/extensions/default.nix:26]: Recently has been documented the review criteria for `vscode-extensions`, you might find it useful:

> https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/editors/vscode/extensions/README.md#conventions-for-adding-new-extensions

Consider:
> Move extension to a discrete directory whenever the extension needs extra parameters/packages (at top of the file) or other files (such as patches, update script, components). Global index file parameters/packages should be utilities shared by many extensions. Extension specific parameters/packages should not be in the global index page.

File: pkgs/applications/editors/vscode/extensions/default.nix [hunk 2]
```diff
@@ -3217,6 +3218,32 @@ let
         meta.license = lib.licenses.lgpl3Only;
       };
 
+      sourcery.sourcery = buildVscodeMarketplaceExtension {
+        mktplcRef = {
+          name = "sourcery";
+          publisher = "sourcery";
+          version = "1.15.0";
+          sha256 = "sha256-aRUg/MZfMj9ZbuGy5CZGZ1LGHC5KMAVYWQAbe0scHYQ=";
```
Comment (superherointj) [inline pkgs/applications/editors/vscode/extensions/default.nix:3226]:
```suggestion
          hash = "sha256-aRUg/MZfMj9ZbuGy5CZGZ1LGHC5KMAVYWQAbe0scHYQ=";
```

File: pkgs/applications/editors/vscode/extensions/default.nix [hunk 3]
```diff
@@ -3217,6 +3218,32 @@ let
...
+          rm -r win ${if stdenv.isLinux then "mac" else "linux"}
+          popd
+        '';
+
+        nativeBuildInputs = lib.optionals stdenv.isLinux [ autoPatchelfHook ];
+        buildInputs = [
+          stdenv.cc.cc.lib
+          libxcrypt-legacy
+        ];
+        meta = {
```
Comment (superherointj) [inline pkgs/applications/editors/vscode/extensions/default.nix:3240]: changelog, downloadPage, homepage is missing?

**Comment by superherointj** [comment]
Please split `vscode-extension` package into it's own commit.

**Comment by superherointj** [comment]
For formatting of **new** Nix files (not for files that existed already, since that would cause lots of changes), I currently recommend using `nixfmt-rfc-style` (available in `nixos-unstable`).

File: pkgs/by-name/so/sourcery/package.nix [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    };
+    "x86_64-darwin" = {
+      platform = "macosx_10_9_universal2";
+      hash = "sha256-XYq49h0KVUGLXeu5mfdlONjpRqmmmp4L5mWTweF/8g8=";
+    };
+  }.${stdenv.system} or (throw "Unsupported platform ${stdenv.system}");
+in
+python3.pkgs.buildPythonApplication rec {
+  pname = "sourcery";
+  version = "1.15.0";
```
Comment (superherointj) [inline pkgs/by-name/so/sourcery/package.nix:23]:
```suggestion
  version = "1.16.0";
```

**Comment by superherointj** [comment]
> I also combined the two commits into one.

It should be a commit per package.

**Comment by superherointj** [comment]
The commit message of Python package is missing the `release` reference. Like:

> https://github.com/sourcery-ai/sourcery/releases/tag/v1.16.0

**Comment by superherointj** [comment]
Also, remember to rebase.

**Comment by TomaSajt** [comment]
I'm not sure if `downloadPage` should be set to what I set it to currently, because you'd probably download it using `pip` or from the `vscode` extension store.

**Comment by TomaSajt** [comment]
Another thing I noticed: this is not the only package with the `sourcery` name.
https://repology.org/projects/?search=sourcery
https://github.com/krzysztofzablocki/Sourcery
It has more stars on GitHub
Should I rename the base-package to something else? (e.g. `sourcery-ai`)

**Comment by superherointj** [comment]
> I'm not sure if `downloadPage` should be set to what I set it to currently, because you'd probably download it using `pip` or from the `vscode` extension store.

- `downloadPage` for the vscode-extension would be link to the vscode marketplace.
- `downloadPage` for the python package would be a link to the pypi page.
Homepage is usually where the source code is kept. (GitHub, GitLab, etc)

**Comment by superherointj** [comment]
The python package should match `pypi` package name.

**Comment by superherointj** [comment]
> I'm not sure if `downloadPage` should be set to what I set it to currently, because you'd probably download it using `pip` or from the `vscode` extension store.

Having this meta information is important to ease upkeep. For example, a maintainer can just click the link to see current situation of package in pypi/vscode marketplace, etc. There are other uses for it. Like the bot will add this information to PR opened. And it can provide meaningful information to anyone looking into the package. In general, we should add it. I don't see a reason to not add it.

**Comment by superherointj** [comment]
Do you still want to call application `sourcery-ai`? You can do that if you wish.

**Comment by TomaSajt** [comment]
> Do you still want to call application `sourcery-ai`? You can do that if you wish.

Well, I don't know what's better. I guess I'll keep it as is.

**Review by superherointj — APPROVED** [review]

**Pull request merged**